pkg/provider/consulcatalog/consul_catalog.go

```diff
@@ -99,20 +113,44 @@ func (p *Provider) Init() error {
 
 // Provide allows the consul catalog provider to provide configurations to traefik using the given configuration channel.
 func (p *Provider) Provide(configurationChan chan<- dynamic.Message, pool *safe.Pool) error {
+	var err error
+	p.client, err = createClient(p.Endpoint)
+	if err != nil {
+		return fmt.Errorf("unable to create consul client: %w", err)
+	}
+
+	if p.ConnectAware {
+		leafWatcher, rootWatcher, err := p.createConnectTLSWatchers()
+		if err != nil {
+			return fmt.Errorf("unable to create consul watch plans: %w", err)
+		}
+		pool.GoCtx(func(routineCtx context.Context) {
+			p.watchConnectTLS(routineCtx, leafWatcher, rootWatcher)
+		})
+	}
+
 	pool.GoCtx(func(routineCtx context.Context) {
 		ctxLog := log.With(routineCtx, log.Str(log.ProviderName, "consulcatalog"))
 		logger := log.FromContext(ctxLog)
 
 		operation := func() error {
-			var err error
-
-			p.client, err = createClient(p.Endpoint)
-			if err != nil {
-				return fmt.Errorf("unable to create consul client: %w", err)
+			var (
+				err      error
+				certInfo *connectCert
+			)
+
+			// If we are running in connect aware mode then we need to
+			// make sure that we obtain the certificates before starting
+			// the service watcher, otherwise a connect enabled service
+			// that gets resolved before the certificates are available
+			// will cause an error condition.
+			if p.ConnectAware {
+				logger.Infof("Waiting for Connect certificate before building first configuration")
+				certInfo = <-p.certChan
 			}
 
 			// get configuration at the provider's startup.
-			err = p.loadConfiguration(routineCtx, configurationChan)
+			err = p.loadConfiguration(routineCtx, certInfo, configurationChan)
```

Comment:
not your fault since it was already like that in the original code, but comparing with the other providers, I think we should also use ctxLog instead of routineCtx here. And same with the other call below L169.